core/ir.mli

```diff
@@ -107,7 +107,7 @@ val binder_of_fun_def : fun_def -> binder
 
 val tapp : value * tyarg list -> value
 
-val letm : binder * tail_computation -> binding
+val letm : ?tyvars:tyvar list -> binder * tail_computation -> binding
```

Comment:
It feels like the only place other than `sugartoir.ml` that calls `letm` is `compilePatterns.ml`. We should not make the `tyvars` optional here, but change the calls  in `compilePatterns.ml` (and verify that creating a let with empty `tyvars` there is actually the right things to do).

Reply:
> verify that creating a let with empty tyvars there is actually the right things to do

I am almost certain it is not the right thing to do - it is what most likely causes all the "unbound type variable" failures in #693 and #694. That being said, I did not yet investigate `compilePatterns`. Indeed, it might turn out that what you're proposing is the next step in fixing the above-mentioned bugs.

Reply:
If investigating `compilePatterns.ml` turns out to be a can of worms, we might just change it to explicitly pass `[]` for now, maintaining the existing behavior and possibly changing that later, if we fix those typing bugs in a dedicated PR.

Reply:
For now I'm inclined not to change `compilePatterns.ml` if there's no good reason to do so. So far it looks like the cause of #693 is somewhere else - when we get to compilation of function into let binding the list of type variables is empty, which clearly is wrong. Once that is fixed there shouldn't be a need to change pattern compilation, I think. But I'd like to move the discussion on that to #693.